In [3]:
import pandas as pd
import numpy as np

from matplotlib import pyplot
pyplot.style.use('fivethirtyeight')

%matplotlib inline

In [4]:
week_1 = pd.read_csv('week1.csv')

In [5]:
week_1.time = pd.to_datetime(week_1.time)

In [27]:
(week_1[(week_1.displayName == 'Matt Ryan') & (week_1.playId == 75)])

time      x      y     s     a   dis       o  \
0   2018-09-07 01:07:14.599000+00:00  91.73  26.67  0.00  0.01  0.02  289.57   
14  2018-09-07 01:07:14.700000+00:00  91.73  26.67  0.00  0.00  0.00  289.57   
28  2018-09-07 01:07:14.799000+00:00  91.73  26.66  0.00  0.00  0.00  288.91   
42  2018-09-07 01:07:14.900000+00:00  91.73  26.66  0.00  0.00  0.01  288.91   
56         2018-09-07 01:07:15+00:00  91.73  26.65  0.00  0.00  0.01  288.28   
70  2018-09-07 01:07:15.099000+00:00  91.73  26.65  0.00  0.00  0.00  287.32   
84  2018-09-07 01:07:15.200000+00:00  91.73  26.66  0.00  0.00  0.01  285.16   
98  2018-09-07 01:07:15.299000+00:00  91.73  26.66  0.00  0.00  0.01  280.45   
112 2018-09-07 01:07:15.400000+00:00  91.73  26.67  0.00  0.00  0.00  277.27   
126 2018-09-07 01:07:15.500000+00:00  91.73  26.67  0.00  0.00  0.01  276.50   
140 2018-09-07 01:07:15.599000+00:00  91.74  26.67  0.00  0.00  0.01  277.41   
154 2018-09-07 01:07:15.700000+00:00  91.76  26.68  0.00  0.00  0.02  282.61   
168 2018-09-07 01:07:15.799000+00:00  91.78  26.68  0.03  0.82  0.02  286.37   
182 2018-09-07 01:07:15.900000+00:00  91.79  26.68  0.22  2.24  0.01  289.31   
196        2018-09-07 01:07:16+00:00  91.84  26.68  0.61  3.46  0.05  292.09   
210 2018-09-07 01:07:16.099000+00:00  91.94  26.67  1.18  4.58  0.10  297.02   
224 2018-09-07 01:07:16.200000+00:00  92.08  26.66  1.67  4.72  0.14  294.38   
238 2018-09-07 01:07:16.299000+00:00  92.28  26.64  2.18  4.51  0.20  290.10   
252 2018-09-07 01:07:16.400000+00:00  92.53  26.61  2.68  4.11  0.25  292.24   
266 2018-09-07 01:07:16.500000+00:00  92.82  26.58  3.13  3.70  0.30  319.46   
280 2018-09-07 01:07:16.599000+00:00  93.15  26.55  3.47  3.02  0.33  342.56   
294 2018-09-07 01:07:16.700000+00:00  93.50  26.51  3.72  2.26  0.36    5.70   
308 2018-09-07 01:07:16.799000+00:00  93.87  26.48  3.81  1.29  0.37   15.96   
322 2018-09-07 01:07:16.900000+00:00  94.25  26.46  3.78  0.46  0.38   21.40   
336        2018-09-07 01:07:17+00:00  94.62  26.44  3.70  0.86  0.37   25.47   
350 2018-09-07 01:07:17.099000+00:00  94.98  26.43  3.53  1.45  0.36   27.43   
364 2018-09-07 01:07:17.200000+00:00  95.31  26.43  3.19  2.28  0.33   30.52   
378 2018-09-07 01:07:17.299000+00:00  95.62  26.43  2.85  2.84  0.30   30.52   
392 2018-09-07 01:07:17.400000+00:00  95.88  26.44  2.47  3.25  0.27   33.20   
406 2018-09-07 01:07:17.500000+00:00  96.10  26.45  2.00  3.65  0.22   36.20   
420 2018-09-07 01:07:17.599000+00:00  96.26  26.48  1.44  4.25  0.16   37.54   
434 2018-09-07 01:07:17.700000+00:00  96.37  26.52  0.98  4.41  0.12   31.26   
448 2018-09-07 01:07:17.799000+00:00  96.43  26.58  0.70  4.20  0.08   26.34   
462 2018-09-07 01:07:17.900000+00:00  96.46  26.64  0.65  3.71  0.06   24.88   
476        2018-09-07 01:07:18+00:00  96.45  26.70  0.76  2.92  0.07    8.35   
490 2018-09-07 01:07:18.099000+00:00  96.43  26.78  0.84  2.05  0.08  356.40   
504 2018-09-07 01:07:18.200000+00:00  96.39  26.85  0.91  1.32  0.08  343.41   
518 2018-09-07 01:07:18.299000+00:00  96.35  26.92  0.89  0.50  0.08  332.18   
532 2018-09-07 01:07:18.400000+00:00  96.31  27.00  0.86  0.11  0.09  323.53   
546 2018-09-07 01:07:18.500000+00:00  96.27  27.07  0.78  1.19  0.08  316.22   
560 2018-09-07 01:07:18.599000+00:00  96.24  27.14  0.69  1.38  0.07  315.32   
574 2018-09-07 01:07:18.700000+00:00  96.24  27.21  0.65  1.46  0.07  315.98   
588 2018-09-07 01:07:18.799000+00:00  96.25  27.27  0.64  1.33  0.06  314.22   
602 2018-09-07 01:07:18.900000+00:00  96.27  27.34  0.64  0.94  0.06  319.15   
616        2018-09-07 01:07:19+00:00  96.29  27.40  0.62  0.50  0.06  323.64   
630 2018-09-07 01:07:19.099000+00:00  96.30  27.46  0.60  0.20  0.06  327.57   
644 2018-09-07 01:07:19.200000+00:00  96.30  27.52  0.59  0.51  0.06  333.07   
658 2018-09-07 01:07:19.299000+00:00  96.29  27.57  0.60  0.87  0.06  335.25   
672 2018-09-07 01:07:19.400000+00:00  96.26  27.64  0.68  1.21  0.07  334.30   
686 2018-09-07 01:07:19.500000+0

In [28]:
week_1[week_1.playId == 75]

time      x      y     s     a   dis       o  \
0   2018-09-07 01:07:14.599000+00:00  91.73  26.67  0.00  0.01  0.02  289.57   
1   2018-09-07 01:07:14.599000+00:00  88.89  36.47  0.01  0.01  0.01  105.63   
2   2018-09-07 01:07:14.599000+00:00  91.35  44.16  0.02  0.03  0.01  290.45   
3   2018-09-07 01:07:14.599000+00:00  86.31  22.01  0.09  0.42  0.01   70.12   
4   2018-09-07 01:07:14.599000+00:00  90.78  36.15  0.00  0.00  0.00  257.61   
..                               ...    ...    ...   ...   ...   ...     ...   
821 2018-09-07 01:07:20.400000+00:00  76.75  52.42  3.86  3.56  0.39   55.22   
822 2018-09-07 01:07:20.400000+00:00  77.41   9.59  3.44  5.45  0.38   96.53   
823 2018-09-07 01:07:20.400000+00:00  78.06  38.54  5.47  3.04  0.56  336.15   
824 2018-09-07 01:07:20.400000+00:00  84.31   5.74  3.78  4.31  0.38   39.11   
825 2018-09-07 01:07:20.400000+00:00  79.37  54.50  5.35  5.82  0.51     NaN   

        dir event      nflId      displayName  jerseyNumber position  frameId  \
0    240.93  None      310.0        Matt Ryan           2.0       QB        1   
1     66.66  None    79848.0  Malcolm Jenkins          27.0       SS        1   
2     16.86  None  2495454.0      Julio Jones          11.0       WR        1   
3    168.91  None  2495613.0     Corey Graham          24.0       FS        1   
4    193.97  None  2533040.0     Mohamed Sanu          12.0       WR        1   
..      ...   ...        ...              ...           ...      ...      ...   
821   22.42  None  2555383.0      Jalen Mills          31.0       CB       59   
822  182.50  None  2555415.0    Austin Hooper          81.0       TE       59   
823  338.39  None  2558175.0       Nate Gerry          47.0       LB       59   
824  246.16  None  2559033.0      Ricky Ortiz          30.0       FB       59   
825     NaN  None        NaN         Football           NaN      NaN       59   

         team      gameId  playId playDirection  route  
0        away  2018090600      75          left    NaN  
1        home  2018090600      75          left    NaN  
2        away  2018090600      75          left  HITCH  
3        home  2018090600      75          left    NaN  
4        away  2018090600      75          left  HITCH  
..        ...         ...     ...           ...    ...  
821      home  2018090600      75          left    NaN  
822      away  2018090600      75          left    OUT  
823      home  2018090600      75          left    NaN  
824      away  2018090600      75          left   FLAT  
825  football  2018090600      75          left    NaN  

[826 rows x 19 columns]

In [53]:
week_1[(week_1.event != 'None') & (week_1.playId == 75)].head(50)

time      x      y      s      a   dis  \
140 2018-09-07 01:07:15.599000+00:00  91.74  26.67   0.00   0.00  0.01   
141 2018-09-07 01:07:15.599000+00:00  88.89  36.50   0.01   0.01  0.00   
142 2018-09-07 01:07:15.599000+00:00  91.36  44.14   0.00   0.01  0.02   
143 2018-09-07 01:07:15.599000+00:00  86.47  22.03   0.18   0.07  0.01   
144 2018-09-07 01:07:15.599000+00:00  90.78  36.16   0.00   0.00  0.00   
145 2018-09-07 01:07:15.599000+00:00  73.61  28.70   0.02   0.02  0.00   
146 2018-09-07 01:07:15.599000+00:00  98.25  26.62   0.00   0.00  0.00   
147 2018-09-07 01:07:15.599000+00:00  86.63  31.12   0.07   0.07  0.03   
148 2018-09-07 01:07:15.599000+00:00  82.63  20.50   0.03   0.03  0.01   
149 2018-09-07 01:07:15.599000+00:00  83.92  43.44   0.06   0.05  0.02   
150 2018-09-07 01:07:15.599000+00:00  91.30  21.85   0.00   0.01  0.00   
151 2018-09-07 01:07:15.599000+00:00  85.70  26.75   0.01   0.02  0.01   
152 2018-09-07 01:07:15.599000+00:00  95.12  26.70   0.00   0.00  0.01   
153 2018-09-07 01:07:15.599000+00:00  90.11  26.85   0.00   0.00  0.00   
490 2018-09-07 01:07:18.099000+00:00  96.43  26.78   0.84   2.05  0.08   
491 2018-09-07 01:07:18.099000+00:00  84.82  38.33   5.53   4.30  0.54   
492 2018-09-07 01:07:18.099000+00:00  80.69  44.91   3.11   6.67  0.34   
493 2018-09-07 01:07:18.099000+00:00  86.02  18.64   3.52   0.69  0.35   
494 2018-09-07 01:07:18.099000+00:00  82.65  34.56   6.35   1.21  0.64   
495 2018-09-07 01:07:18.099000+00:00  69.63  30.81   2.93   2.65  0.30   
496 2018-09-07 01:07:18.099000+00:00  93.81  27.30   3.92   3.23  0.39   
497 2018-09-07 01:07:18.099000+00:00  86.10  33.53   3.75   1.49  0.37   
498 2018-09-07 01:07:18.099000+00:00  79.51  20.00   2.99   0.69  0.30   
499 2018-09-07 01:07:18.099000+00:00  76.53  44.93   4.68   2.18  0.47   
500 2018-09-07 01:07:18.099000+00:00  83.20  21.36   6.21   2.68  0.62   
501 2018-09-07 01:07:18.099000+00:00  83.06  25.98   4.08   3.73  0.40   
502 2018-09-07 01:07:18.099000+00:00  89.62  18.39   6.75   1.51  0.67   
503 2018-09-07 01:07:18.099000+00:00  96.90  26.97   2.05   3.76  0.22   
644 2018-09-07 01:07:19.200000+00:00  96.30  27.52   0.59   0.51  0.06   
645 2018-09-07 01:07:19.200000+00:00  82.34  45.18   7.57   1.72  0.75   
646 2018-09-07 01:07:19.200000+00:00  81.11  47.87   4.23   1.63  0.42   
647 2018-09-07 01:07:19.200000+00:00  82.90  17.62   3.64   4.01  0.36   
648 2018-09-07 01:07:19.200000+00:00  77.66  35.52   2.99   3.53  0.31   
649 2018-09-07 01:07:19.200000+00:00  70.28  35.09   6.10   2.74  0.59   
650 2018-09-07 01:07:19.200000+00:00  91.13  30.53   3.85   2.91  0.38   
651 2018-09-07 01:07:19.200000+00:00  82.76  36.69   4.85   2.33  0.48   
652 2018-09-07 01:07:19.200000+00:00  77.11  17.30   4.83   3.10  0.47   
653 2018-09-07 01:07:19.200000+00:00  74.40  47.11   4.05   4.64  0.38   
654 2018-09-07 01:07:19.200000+00:00  78.35  16.74   7.05   4.26  0.70   
655 2018-09-07 01:07:19.200000+00:00  79.94  31.07   6.47   2.47  0.64   
656 2018-09-07 01:07:19.200000+00:00  87.13  10.93   6.81   2.09  0.69   
657 2018-09-07 01:07:19.200000+00:00  83.29  46.47  12.68  16.51  1.90   
714 2018-09-07 01:07:19.700000+00:00  96.10  27.82   0.99   1.14  0.10   
715 2018-09-07 01:07:19.700000+00:00  81.74  49.08   7.94   2.37  0.80   
716 2018-09-07 01:07:19.700000+00:00  81.48  50.13   4.93   3.07  0.49   
717 2018-09-07 01:07:19.700000+00:00  81.41  19.08   4.65   2.84  0.46   
718 2018-09-07 01:07:19.700000+00:00  76.75  36.60   2.82   1.85  0.28   
719 2018-09-07 01:07:19.700000+00:00  71.81  37.93   6.57   1.61  0.66   
720 2018-09-07 01:07:19.700000+00:00  89.30  30.99   3.60   1.45  0.37   
721 2018-09-07 01:07:19.700000+00:00  81.51  39.01   5.58   0.92  0.56   

          o     dir                event      nflId      displayName  \
140  277.41  235.01            ball_snap      310.0        Matt Ryan   
141  104.49   36.45            ball_snap    79848.0  Malcolm Jenkins   
142  289.53   49.86            ball_snap 

In [51]:
week_1[(week_1.event != 'None') & (week_1.playId == 75)].tail(20)

time      x      y     s      a   dis       o  \
722 2018-09-07 01:07:19.700000+00:00  76.54  14.86  4.91   1.36  0.49  249.15   
723 2018-09-07 01:07:19.700000+00:00  75.30  49.33  5.31   1.22  0.53   57.98   
724 2018-09-07 01:07:19.700000+00:00  77.70  13.31  6.73   2.22  0.68  137.26   
725 2018-09-07 01:07:19.700000+00:00  79.25  34.33  6.73   0.58  0.67  356.66   
726 2018-09-07 01:07:19.700000+00:00  86.28   8.01  5.34   3.77  0.55   71.65   
727 2018-09-07 01:07:19.700000+00:00  81.52  50.88  5.50  10.01  0.60     NaN   
798 2018-09-07 01:07:20.299000+00:00  95.47  28.23  1.42   0.41  0.14  297.30   
799 2018-09-07 01:07:20.299000+00:00  80.12  53.23  6.83   3.50  0.70  340.81   
800 2018-09-07 01:07:20.299000+00:00  80.09  54.09  7.26   2.23  0.73  334.19   
801 2018-09-07 01:07:20.299000+00:00  80.14  21.60  4.44   1.80  0.45  340.64   
802 2018-09-07 01:07:20.299000+00:00  76.25  38.20  2.59   1.44  0.26  352.40   
803 2018-09-07 01:07:20.299000+00:00  72.89  41.43  5.48   3.10  0.56   22.15   
804 2018-09-07 01:07:20.299000+00:00  87.52  31.32  2.41   2.11  0.25  325.31   
805 2018-09-07 01:07:20.299000+00:00  80.11  41.98  5.06   2.06  0.52  331.93   
806 2018-09-07 01:07:20.299000+00:00  75.51  12.77  2.88   4.19  0.30  300.98   
807 2018-09-07 01:07:20.299000+00:00  76.60  52.06  4.52   2.37  0.46   55.22   
808 2018-09-07 01:07:20.299000+00:00  77.43   9.96  4.10   4.96  0.44  104.54   
809 2018-09-07 01:07:20.299000+00:00  78.26  38.01  5.78   2.70  0.59  336.73   
810 2018-09-07 01:07:20.299000+00:00  84.65   5.91  3.87   4.57  0.39   42.58   
811 2018-09-07 01:07:20.299000+00:00  79.76  54.17  5.85   4.36  0.40     NaN   

        dir                event      nflId      displayName  jerseyNumber  \
722  195.28  pass_outcome_caught  2552689.0     Ronald Darby          21.0   
723   25.69  pass_outcome_caught  2555383.0      Jalen Mills          31.0   
724  186.71  pass_outcome_caught  2555415.0    Austin Hooper          81.0   
725  348.67  pass_outcome_caught  2558175.0       Nate Gerry          47.0   
726  202.39  pass_outcome_caught  2559033.0      Ricky Ortiz          30.0   
727     NaN  pass_outcome_caught        NaN         Football           NaN   
798  297.36        out_of_bounds      310.0        Matt Ryan           2.0   
799  330.77        out_of_bounds    79848.0  Malcolm Jenkins          27.0   
800  334.41        out_of_bounds  2495454.0      Julio Jones          11.0   
801  338.52        out_of_bounds  2495613.0     Corey Graham          24.0   
802  351.16        out_of_bounds  2533040.0     Mohamed Sanu          12.0   
803   10.18        out_of_bounds  2534832.0    Rodney McLeod          23.0   
804  286.01        out_of_bounds  2543583.0  Devonta Freeman          24.0   
805  332.83        out_of_bounds  2552315.0     Jordan Hicks          58.0   
806  221.69        out_of_bounds  2552689.0     Ronald Darby          21.0   
807   23.06        out_of_bounds  2555383.0      Jalen Mills          31.0   
808  183.45        out_of_bounds  2555415.0    Austin Hooper          81.0   
809  340.55        out_of_bounds  2558175.0       Nate Gerry          47.0   
810  238.71        out_of_bounds  2559033.0      Ricky Ortiz          30.0   
811     NaN        out_of_bounds        NaN         Football           NaN   

    position  frameId      team      gameId  playId playDirection  route  
722       CB       52      home  2018090600      75          left    NaN  
723       CB       52      home  2018090600      75          left    NaN  
724       TE       52      away  2018090600      75          left    OUT  
725       LB       52      home  2018090600      75          left    NaN  
726       FB       52      away  2018090600      75          left   FLAT  
727      NaN       52  football  2018090600      75          left    NaN  
798       QB       58      away  2018090600      75          left    NaN  
799       SS       58      home  2018090600      75          left    NaN  
800       WR    

In [55]:
week_1[(week_1.event != 'None') & (week_1.playId == 75) & 
       ((week_1.displayName == 'Julio Jones') | (week_1.displayName == 'Jalen Mills'))]

time      x      y     s     a   dis       o  \
142 2018-09-07 01:07:15.599000+00:00  91.36  44.14  0.00  0.01  0.02  289.53   
149 2018-09-07 01:07:15.599000+00:00  83.92  43.44  0.06  0.05  0.02   99.77   
492 2018-09-07 01:07:18.099000+00:00  80.69  44.91  3.11  6.67  0.34  155.11   
499 2018-09-07 01:07:18.099000+00:00  76.53  44.93  4.68  2.18  0.47  182.14   
646 2018-09-07 01:07:19.200000+00:00  81.11  47.87  4.23  1.63  0.42   53.31   
653 2018-09-07 01:07:19.200000+00:00  74.40  47.11  4.05  4.64  0.38   69.31   
716 2018-09-07 01:07:19.700000+00:00  81.48  50.13  4.93  3.07  0.49  326.77   
723 2018-09-07 01:07:19.700000+00:00  75.30  49.33  5.31  1.22  0.53   57.98   
800 2018-09-07 01:07:20.299000+00:00  80.09  54.09  7.26  2.23  0.73  334.19   
807 2018-09-07 01:07:20.299000+00:00  76.60  52.06  4.52  2.37  0.46   55.22   

        dir                event      nflId  displayName  jerseyNumber  \
142   49.86            ball_snap  2495454.0  Julio Jones          11.0   
149  228.90            ball_snap  2555383.0  Jalen Mills          31.0   
492  297.88         pass_forward  2495454.0  Julio Jones          11.0   
499  277.48         pass_forward  2555383.0  Jalen Mills          31.0   
646   22.58         pass_arrived  2495454.0  Julio Jones          11.0   
653   13.45         pass_arrived  2555383.0  Jalen Mills          31.0   
716  358.86  pass_outcome_caught  2495454.0  Julio Jones          11.0   
723   25.69  pass_outcome_caught  2555383.0  Jalen Mills          31.0   
800  334.41        out_of_bounds  2495454.0  Julio Jones          11.0   
807   23.06        out_of_bounds  2555383.0  Jalen Mills          31.0   

    position  frameId  team      gameId  playId playDirection  route  
142       WR       11  away  2018090600      75          left  HITCH  
149       CB       11  home  2018090600      75          left    NaN  
492       WR       36  away  2018090600      75          left  HITCH  
499       CB       36  home  2018090600      75          left    NaN  
646       WR       47  away  2018090600      75          left  HITCH  
653       CB       47  home  2018090600      75          left    NaN  
716       WR       52  away  2018090600      75          left  HITCH  
723       CB       52  home  2018090600      75          left    NaN  
800       WR       58  away  2018090600      75          left  HITCH  
807       CB       58  home  2018090600      75          left    NaN

In [56]:
week_1[(week_1.event != 'None') & (week_1.playId == 75) & 
       ((week_1.displayName == 'Mohamed Sanu') | (week_1.displayName == 'Malcolm Jenkins'))]

time      x      y     s     a   dis       o  \
141 2018-09-07 01:07:15.599000+00:00  88.89  36.50  0.01  0.01  0.00  104.49   
144 2018-09-07 01:07:15.599000+00:00  90.78  36.16  0.00  0.00  0.00  259.19   
491 2018-09-07 01:07:18.099000+00:00  84.82  38.33  5.53  4.30  0.54  295.34   
494 2018-09-07 01:07:18.099000+00:00  82.65  34.56  6.35  1.21  0.64  262.06   
645 2018-09-07 01:07:19.200000+00:00  82.34  45.18  7.57  1.72  0.75  354.68   
648 2018-09-07 01:07:19.200000+00:00  77.66  35.52  2.99  3.53  0.31   78.88   
715 2018-09-07 01:07:19.700000+00:00  81.74  49.08  7.94  2.37  0.80    0.01   
718 2018-09-07 01:07:19.700000+00:00  76.75  36.60  2.82  1.85  0.28   33.31   
799 2018-09-07 01:07:20.299000+00:00  80.12  53.23  6.83  3.50  0.70  340.81   
802 2018-09-07 01:07:20.299000+00:00  76.25  38.20  2.59  1.44  0.26  352.40   

        dir                event      nflId      displayName  jerseyNumber  \
141   36.45            ball_snap    79848.0  Malcolm Jenkins          27.0   
144  294.93            ball_snap  2533040.0     Mohamed Sanu          12.0   
491  318.58         pass_forward    79848.0  Malcolm Jenkins          27.0   
494  272.78         pass_forward  2533040.0     Mohamed Sanu          12.0   
645  353.52         pass_arrived    79848.0  Malcolm Jenkins          27.0   
648  305.48         pass_arrived  2533040.0     Mohamed Sanu          12.0   
715  347.39  pass_outcome_caught    79848.0  Malcolm Jenkins          27.0   
718  332.72  pass_outcome_caught  2533040.0     Mohamed Sanu          12.0   
799  330.77        out_of_bounds    79848.0  Malcolm Jenkins          27.0   
802  351.16        out_of_bounds  2533040.0     Mohamed Sanu          12.0   

    position  frameId  team      gameId  playId playDirection  route  
141       SS       11  home  2018090600      75          left    NaN  
144       WR       11  away  2018090600      75          left  HITCH  
491       SS       36  home  2018090600      75          left    NaN  
494       WR       36  away  2018090600      75          left  HITCH  
645       SS       47  home  2018090600      75          left    NaN  
648       WR       47  away  2018090600      75          left  HITCH  
715       SS       52  home  2018090600      75          left    NaN  
718       WR       52  away  2018090600      75          left  HITCH  
799       SS       58  home  2018090600      75          left    NaN  
802       WR       58  away  2018090600      75          left  HITCH

The data is organized on a play by play basis. So for each play it will give the positional data of every single player on that play involved in it. This will likely need to be tied into the plays dataframe also given. 

PlayId == 75 was a 15 yard pass play to julio jones. So using that data and this one we can see that julio jones ran a hitch route where the player makes a hard sprint then turns and stops. So we can see based on the y location that julio jones was being guarded by jalen mills. And jalen mills gave him an 8 yard cushion on the snap. Then right as the pass was being thrown he had an acceleration of 6.67 yards/s^2 where jalen mills waiting for him had a much slower speed of 2.18. Then as the ball is arriving we see that jalen mills begins to speed up to match it where as julio, running a hitch slows down instead. And creates more distance when the pass arrives then when it is thrown. 

These are factors that will weigh heavily on a model to see if a pass is completed or not. For example on a hitch route taking data like the speed of the players and the distance between them on the pass forward and the pass arrival are very important as both give a good idea of how much room the receiver has to make the completion. 

Even on this play Sanu, running the same route does not get the ball and when you look at his acceleration/speed and distance to his defender malcolm jenkins you can see why. Jenkins is all over him on the pass forward and moving at a faster acceleration. So having less space will mean he is likely not to get the ball. If Julio jones was covered the same way as the only 2 real passing threats this will have likely led to an incomplete pass or even possibly a sack. So those are the factors that will need to be looked at further. 

My next step will be to look at some bad plays, sacks and incompletions. And see what similarities we can find in those actions whether its good defense or maybe just bad throws. 


# Looking at a play that results in a sack

In [2]:
plays_df = pd.read_csv('plays.csv')

In [11]:
plays_df.head(14)

gameId  playId                                    playDescription  \
0   2018090600      75  (15:00) M.Ryan pass short right to J.Jones pus...   
1   2018090600     146  (13:10) M.Ryan pass incomplete short right to ...   
2   2018090600     168  (13:05) (Shotgun) M.Ryan pass incomplete short...   
3   2018090600     190  (13:01) (Shotgun) M.Ryan pass deep left to J.J...   
4   2018090600     256  (10:59) (Shotgun) M.Ryan pass incomplete short...   
5   2018090600     320  (10:10) (Shotgun) N.Foles pass short left to N...   
6   2018090600     344  (9:24) (Shotgun) N.Foles pass incomplete short...   
7   2018090600     402  (9:08) M.Ryan pass incomplete deep left to M.S...   
8   2018090600     492  (7:01) M.Ryan pass short left to T.Coleman pus...   
9   2018090600     521  (6:19) M.Ryan pass short left to A.Hooper to P...   
10  2018090600     545          (5:38) M.Ryan pass incomplete short left.   
11  2018090600     636  (5:07) (Shotgun) M.Ryan pass incomplete short ...   
12  2018090600     752  (3:46) (Shotgun) N.Foles pass short left to N....   
13  2018090600     776  (3:06) (Shotgun) N.Foles sacked at PHI 36 for ...   

    quarter  down  yardsToGo possessionTeam        playType yardlineSide  \
0         1     1         15            ATL  play_type_pass          ATL   
1         1     1         10            ATL  play_type_pass          PHI   
2         1     2         10            ATL  play_type_pass          PHI   
3         1     3         10            ATL  play_type_pass          PHI   
4         1     3          1            ATL  play_type_pass          PHI   
5         1     2          8            PHI  play_type_pass          PHI   
6         1     3          4            PHI  play_type_pass          PHI   
7         1     1         10            ATL  play_type_pass          PHI   
8         1     2         13            ATL  play_type_pass          PHI   
9         1     1         10            ATL  play_type_pass          PHI   
10        1     2          7            ATL  play_type_pass          PHI   
11        1     3          3            ATL  play_type_pass          PHI   
12        1     2          4            PHI  play_type_pass          PHI   
13        1     1         10            PHI  play_type_sack          PHI   

    yardlineNumber  ... preSnapHomeScore gameClock  absoluteYardlineNumber  \
0               20  ...              0.0  15:00:00                    90.0   
1               39  ...              0.0  13:10:00                    49.0   
2               39  ...              0.0  13:05:00                    49.0   
3               39  ...              0.0  13:01:00                    49.0   
4                1  ...              0.0  10:59:00                    11.0   
5                4  ...              0.0  10:10:00                    14.0   
6                8  ...              0.0  09:24:00                    18.0   
7               44  ...              0.0  09:08:00                    54.0   
8               36  ...              0.0  07:01:00                    46.0   
9               10  ...              0.0  06:19:00                    20.0   
10               7  ...              0.0  05:38:00                    17.0   
11               3  ...              0.0  05:07:00                    13.0   
12              37  ...              0.0  03:46:00                    47.0   
13              43  ...              0.0  03:06:00                    53.0   

    penaltyCodes penaltyJerseyNumbers passResult  offensePlayResult  \
0            NaN                  NaN          C                 10   
1            NaN                  NaN          I                  0   
2            NaN                  NaN          I                  0   
3            NaN                  NaN          C                 33   
4            NaN                  NaN          I                  0   
5            NaN                  NaN          C                  4   
6            NaN                  NaN      

In [ ]:
# play 776

In [10]:
week_1[(week_1.playId == 776) & (week_1.event != 'None') & (week_1.position == 'WR')]

time      x      y     s     a   dis  \
11249  2018-09-07 01:37:30.799000+00:00  47.45  27.73  0.00  0.00  0.01   
11819  2018-09-07 01:37:34.599000+00:00  48.52  29.04  1.58  0.63  0.16   
161030 2018-09-09 17:30:08.599000+00:00  81.85  22.01  0.02  0.02  0.01   
161282 2018-09-09 17:30:10.400000+00:00  82.05  21.91  1.16  2.95  0.11   
161492 2018-09-09 17:30:11.900000+00:00  79.45  23.21  0.97  2.43  0.11   
161576 2018-09-09 17:30:12.500000+00:00  79.38  23.35  0.36  1.42  0.03   
161604 2018-09-09 17:30:12.700000+00:00  79.47  23.30  0.61  0.99  0.06   

             o     dir                event      nflId     displayName  \
11249   125.68  166.45            ball_snap  2506467.0  Darren Sproles   
11819   263.54  231.21              qb_sack  2506467.0  Darren Sproles   
161030  104.22   48.74            ball_snap  2556466.0   Peyton Barber   
161282  150.91  284.15         pass_forward  2556466.0   Peyton Barber   
161492  230.29  322.94         pass_arrived  2556466.0   Peyton Barber   
161576  184.41  117.28  pass_outcome_caught  2556466.0   Peyton Barber   
161604  163.71  121.99               tackle  2556466.0   Peyton Barber   

        jerseyNumber position  frameId  team      gameId  playId  \
11249           43.0       RB       11  home  2018090600     776   
11819           43.0       RB       49  home  2018090600     776   
161030          25.0       RB       11  away  2018090906     776   
161282          25.0       RB       29  away  2018090906     776   
161492          25.0       RB       44  away  2018090906     776   
161576          25.0       RB       50  away  2018090906     776   
161604          25.0       RB       52  away  2018090906     776   

       playDirection route  
11249          right   NaN  
11819          right   NaN  
161030         right   NaN  
161282         right   NaN  
161492         right   NaN  
161576         right   NaN  
161604         right   NaN

In [9]:
week_1[(week_1.playId == 776) & (week_1.event != 'None') & (week_1.team == 'away')]

time       x      y     s     a   dis  \
11252  2018-09-07 01:37:30.799000+00:00   56.26   6.92  0.00  0.00  0.01   
11253  2018-09-07 01:37:30.799000+00:00   53.73  36.87  0.00  0.00  0.00   
11255  2018-09-07 01:37:30.799000+00:00   56.89  16.74  0.00  0.00  0.00   
11256  2018-09-07 01:37:30.799000+00:00   53.61  34.88  0.00  0.00  0.00   
11259  2018-09-07 01:37:30.799000+00:00   57.45  28.69  0.00  0.00  0.00   
11260  2018-09-07 01:37:30.799000+00:00   53.41  24.25  0.03  0.04  0.03   
11261  2018-09-07 01:37:30.799000+00:00   68.77  26.34  0.09  0.08  0.03   
11262  2018-09-07 01:37:30.799000+00:00   53.38  22.02  0.80  1.52  0.07   
11822  2018-09-07 01:37:34.599000+00:00   75.23  11.59  6.05  3.37  0.62   
11823  2018-09-07 01:37:34.599000+00:00   64.76  41.69  1.18  5.73  0.12   
11825  2018-09-07 01:37:34.599000+00:00   64.68  17.85  1.18  3.21  0.10   
11826  2018-09-07 01:37:34.599000+00:00   56.70  33.97  3.11  2.34  0.30   
11829  2018-09-07 01:37:34.599000+00:00   73.71  31.35  8.14  2.28  0.82   
11830  2018-09-07 01:37:34.599000+00:00   45.99  28.32  2.83  0.57  0.28   
11831  2018-09-07 01:37:34.599000+00:00   79.72  28.99  5.26  1.62  0.53   
11832  2018-09-07 01:37:34.599000+00:00   42.79  32.39  2.83  0.23  0.27   
161023 2018-09-09 17:30:08.599000+00:00   84.50  23.78  0.05  0.03  0.02   
161026 2018-09-09 17:30:08.599000+00:00   84.25   8.94  0.01  0.01  0.01   
161027 2018-09-09 17:30:08.599000+00:00   84.87  19.24  0.07  0.06  0.01   
161028 2018-09-09 17:30:08.599000+00:00   84.04  35.12  0.01  0.01  0.01   
161030 2018-09-09 17:30:08.599000+00:00   81.85  22.01  0.02  0.02  0.01   
161033 2018-09-09 17:30:08.599000+00:00   84.81  43.84  0.02  0.02  0.00   
161275 2018-09-09 17:30:10.400000+00:00   81.57  23.97  1.45  2.04  0.15   
161278 2018-09-09 17:30:10.400000+00:00   88.94   7.30  5.88  1.79  0.59   
161279 2018-09-09 17:30:10.400000+00:00   88.60  18.20  4.25  2.36  0.42   
161280 2018-09-09 17:30:10.400000+00:00   87.72  37.49  4.92  3.23  0.49   
161282 2018-09-09 17:30:10.400000+00:00   82.05  21.91  1.16  2.95  0.11   
161285 2018-09-09 17:30:10.400000+00:00   89.47  44.84  5.62  2.07  0.55   
161485 2018-09-09 17:30:11.900000+00:00   77.64  25.28  2.22  4.44  0.24   
161488 2018-09-09 17:30:11.900000+00:00   99.55   4.60  7.62  2.57  0.77   
161489 2018-09-09 17:30:11.900000+00:00   89.87  12.23  4.73  1.45  0.48   
161490 2018-09-09 17:30:11.900000+00:00   88.60  45.61  4.67  3.05  0.48   
161492 2018-09-09 17:30:11.900000+00:00   79.45  23.21  0.97  2.43  0.11   
161495 2018-09-09 17:30:11.900000+00:00   98.39  46.88  5.50  2.85  0.56   
161569 2018-09-09 17:30:12.500000+00:00   77.32  25.44  0.57  2.55  0.04   
161572 2018-09-09 17:30:12.500000+00:00  103.56   2.83  7.05  2.27  0.71   
161573 2018-09-09 17:30:12.500000+00:00   89.11   9.78  3.76  1.91  0.38   
161574 2018-09-09 17:30:12.500000+00:00   88.12  47.71  2.60  3.06  0.27   
161576 2018-09-09 17:30:12.500000+00:00   79.38  23.35  0.36  1.42  0.03   
161579 2018-09-09 17:30:12.500000+00:00  101.28  45.92  4.76  3.27  0.48   
161597 2018-09-09 17:30:12.700000+00:00   77.43  25.40  0.72  0.94  0.07   
161600 2018-09-09 17:30:12.700000+00:00  104.73   2.13  6.61  2.38  0.67   
161601 2018-09-09 17:30:12.700000+00:00   88.94   9.09  3.33  2.22  0.34   
161602 2018-09-09 17:30:12.700000+00:00   87.92  48.11  1.97  2.93  0.21   
161604 2018-09-09 17:30:12.700000+00:00   79.47  23.30  0.61  0.99  0.06   
161607 2018-09-09 17:30:12.700000+00:00  102.04  45.38  4.59  3.16  0.46   

             o     dir                event      nflId         displayName  \
11252   349.87   67.35            ball_snap  2539334.0     Desmond Trufant   
11253   234.18  340.50            ball_snap  2539653.0       Robert Alford   
11255   262.05   54.21            ball_snap  2543850.0       Ricardo Allen   
11256   251.00  321.71            ball_snap  2552301.0         Vic Beasley   
11259   274.11    0.40            ball_snap  2555162.0         Deion Jon

In [ ]:
# desmond trufant & deAndre carter
# Brian poole & Mike wallace
# Ricardo allen & Nelson agholor

In [12]:
week_1[(week_1.playId == 776) & (week_1.event != 'None') & 
       ((week_1.displayName == 'Desmond Trufant') | (week_1.displayName == 'DeAndre Carter'))]

time      x      y     s     a   dis  \
11252 2018-09-07 01:37:30.799000+00:00  56.26   6.92  0.00  0.00  0.01   
11258 2018-09-07 01:37:30.799000+00:00  51.45   6.86  0.00  0.00  0.00   
11822 2018-09-07 01:37:34.599000+00:00  75.23  11.59  6.05  3.37  0.62   
11828 2018-09-07 01:37:34.599000+00:00  59.40  13.36  2.58  2.58  0.27   

            o     dir      event      nflId      displayName  jerseyNumber  \
11252  349.87   67.35  ball_snap  2539334.0  Desmond Trufant          21.0   
11258   85.86  117.88  ball_snap  2553502.0   DeAndre Carter          16.0   
11822   21.88   65.83    qb_sack  2539334.0  Desmond Trufant          21.0   
11828  334.35  332.14    qb_sack  2553502.0   DeAndre Carter          16.0   

      position  frameId  team      gameId  playId playDirection route  
11252       CB       11  away  2018090600     776         right   NaN  
11258       WR       11  home  2018090600     776         right   NaN  
11822       CB       49  away  2018090600     776         right   NaN  
11828       WR       49  home  2018090600     776         right   NaN

In [13]:
week_1[(week_1.playId == 776) & (week_1.event != 'None') & 
       ((week_1.displayName == 'Brian Poole') | (week_1.displayName == 'Mike Wallace'))]

time      x      y     s     a   dis  \
11250 2018-09-07 01:37:30.799000+00:00  51.05  21.59  0.00  0.00  0.00   
11262 2018-09-07 01:37:30.799000+00:00  53.38  22.02  0.80  1.52  0.07   
11820 2018-09-07 01:37:34.599000+00:00  74.54  30.39  9.01  2.31  0.91   
11832 2018-09-07 01:37:34.599000+00:00  42.79  32.39  2.83  0.23  0.27   

            o     dir      event      nflId   displayName  jerseyNumber  \
11250   95.65   81.79  ball_snap  2507763.0  Mike Wallace          14.0   
11262  269.07  341.89  ball_snap  2556445.0   Brian Poole          34.0   
11820   19.67   45.32    qb_sack  2507763.0  Mike Wallace          14.0   
11832   90.16  286.22    qb_sack  2556445.0   Brian Poole          34.0   

      position  frameId  team      gameId  playId playDirection route  
11250       WR       11  home  2018090600     776         right   NaN  
11262       CB       11  away  2018090600     776         right   NaN  
11820       WR       49  home  2018090600     776         right   NaN  
11832       CB       49  away  2018090600     776         right   NaN

In [14]:
week_1[(week_1.playId == 776) & (week_1.event != 'None') & 
       ((week_1.displayName == 'Ricardo Allen') | (week_1.displayName == 'Nelson Agholor'))]

time      x      y     s     a   dis  \
11255 2018-09-07 01:37:30.799000+00:00  56.89  16.74  0.00  0.00  0.00   
11257 2018-09-07 01:37:30.799000+00:00  50.32  16.35  0.00  0.00  0.00   
11825 2018-09-07 01:37:34.599000+00:00  64.68  17.85  1.18  3.21  0.10   
11827 2018-09-07 01:37:34.599000+00:00  69.95  21.82  6.19  4.35  0.62   

            o     dir      event      nflId     displayName  jerseyNumber  \
11255  262.05   54.21  ball_snap  2543850.0   Ricardo Allen          37.0   
11257   79.45  124.61  ball_snap  2552600.0  Nelson Agholor          13.0   
11825  299.02  270.58    qb_sack  2543850.0   Ricardo Allen          37.0   
11827  342.65   28.71    qb_sack  2552600.0  Nelson Agholor          13.0   

      position  frameId  team      gameId  playId playDirection route  
11255       SS       11  away  2018090600     776         right   NaN  
11257       WR       11  home  2018090600     776         right   NaN  
11825       SS       49  away  2018090600     776         right   NaN  
11827       WR       49  home  2018090600     776         right   NaN

So we can see here that unlike the previous play there is less space between the offensive and the defensive players here than there was in the play when there was a complete pass. In the julio jones play previously there was 8 yards between the offensive and defensive player when the pass was sent. In this play the most distance between the offensive and defensive player was 4 yards. and in that play the acceleration of the defensive player was similar to the offensive one so they could have likely closed the space. 

So just from an initial look we can see on these 2 plays that when the space and speed between the offensive and defensive players is smaller the qb has less room to throw and it will likely result in a less productive possession. in this case a sack.

Next step will be to take a look at more plays like this then come up with other factors that could determine that could be used to determine good and bad plays. Play types, time between the snap and the throw, personnel, etc.